### ISSUE TITLE
`break` without labels should be hygienic

### ISSUE THREAD

**Issue by lfairy** [rust-lang/rust#38663]
Minimal test case:

```rust
macro_rules! blah {
    ($expr:expr) => {
        loop {
            $expr;
            println!("reached end of macro loop");
            break;
        }
    };
}

fn main() {
    loop {
        blah!(break);
        println!("reached end of main loop");
        break;
    }
    println!("reached end of program");
}
```

Now, since the call `blah!(break)` is written in the context of `main()`, I would expect that `break` to exit the *outer* loop. That is,

```
$ rustc test.rs && ./test
reached end of program
```

But instead, the `break` is hooked up to the loop within the macro:

```
$ rustc test.rs && ./test
reached end of main loop
reached end of program
```

This behavior seems unintuitive, especially when the equivalent code with labels

```rust
macro_rules! blah {
    ($expr:expr) => {
        'a: loop {
            $expr;
            println!("reached end of macro loop");
            break 'a;
        }
    };
}

fn main() {
    'a: loop {
        blah!(break 'a);
        println!("reached end of main loop");
        break;
    }
    println!("reached end of program");
}
```

*does* respect hygiene.

**Comment by durka** [comment]
The second case is definitely working as expected due to identifier hygiene (it prints a warning which seems incorrect though).

The first case is somewhat unintuitive, I agree, but can't be changed for stability (at least until macros 2.0, but I'm not sure it is a good idea even then to change subtle things like this out of concern for porting).

**Comment by DemiMarie** [comment]
In that case, I think we should issue a warning.

**Comment by petrochenkov** [comment]
It'd be strange to make `break` hygienic without having all the remaining control flow constructions hygienic in general (`return`, `else`?), it's a large and very breaking change, in case of `return` at least.

**Comment by petrochenkov** [comment]
This is a sufficiently large and breaking change and it requires an RFC.
I'm going to close this issue, but feel free to reopen it in [the RFC repo](https://github.com/rust-lang/rfcs) if you are going to pursue this.
